packages/open-next/src/adapters/util.ts

```diff
@@ -1,66 +1,103 @@
   const { config } = JSON.parse(json);
   return config as NextConfig;
 }
 export function loadBuildId(nextDir: string) {
-  const filePath = path.join(nextDir, "BUILD_ID");
-  return fs.readFileSync(filePath, "utf-8").trim();
+  const filePath = path.join(nextDir, 'BUILD_ID');
+  return fs.readFileSync(filePath, 'utf-8').trim();
 }
 
 export function loadHtmlPages(nextDir: string) {
-  const filePath = path.join(nextDir, "server", "pages-manifest.json");
-  const json = fs.readFileSync(filePath, "utf-8");
+  const filePath = path.join(nextDir, 'server', 'pages-manifest.json');
+  const json = fs.readFileSync(filePath, 'utf-8');
   return Object.entries(JSON.parse(json))
-    .filter(([_, value]) => (value as string).endsWith(".html"))
+    .filter(([_, value]) => (value as string).endsWith('.html'))
     .map(([key]) => key);
 }
 
 export function loadPublicAssets(openNextDir: string) {
-  const filePath = path.join(openNextDir, "public-files.json");
-  const json = fs.readFileSync(filePath, "utf-8");
+  const filePath = path.join(openNextDir, 'public-files.json');
+  const json = fs.readFileSync(filePath, 'utf-8');
   return JSON.parse(json) as PublicFiles;
 }
 
 export function loadRoutesManifest(nextDir: string) {
-  const filePath = path.join(nextDir, "routes-manifest.json");
-  const json = fs.readFileSync(filePath, "utf-8");
+  const filePath = path.join(nextDir, 'routes-manifest.json');
+  const json = fs.readFileSync(filePath, 'utf-8');
   const routesManifest = JSON.parse(json) as RoutesManifest;
-  // Static routes take precedence over dynamic routes
-  return [...routesManifest.staticRoutes, ...routesManifest.dynamicRoutes];
+
+  // TODO: do we need to include rewrites.afterFiles and rewrites.fallback here?
+  const rewrites = routesManifest.rewrites.beforeFiles ?? [];
+  // Rewrite routes first, then static routes then dynamic routes
+  return {
+    rewrites,
+    redirects: routesManifest.redirects,
+    routes: [...routesManifest.staticRoutes, ...routesManifest.dynamicRoutes],
+  };
 }
 
 export function loadAppPathsManifestKeys(nextDir: string) {
   const appPathsManifestPath = path.join(
     nextDir,
-    "server",
-    "app-paths-manifest.json"
+    'server',
+    'app-paths-manifest.json'
   );
   const appPathsManifestJson = fs.existsSync(appPathsManifestPath)
-    ? fs.readFileSync(appPathsManifestPath, "utf-8")
-    : "{}";
+    ? fs.readFileSync(appPathsManifestPath, 'utf-8')
+    : '{}';
   const appPathsManifest = JSON.parse(appPathsManifestJson) as Record<
     string,
     string
   >;
   return Object.keys(appPathsManifest).map((key) => {
-    // Remove group route params and /page suffix
-    const cleanedKey = key.replace(/\/\(\w+\)|\/page$/g, "");
+    // Remove parallel route
+    let cleanedKey = key.replace(/\/@[^\/]+/g, '');
+
+    // Remove group routes
+    cleanedKey = cleanedKey.replace(/\/\((?!\.)[^\)]*\)/g, '');
+
+    // Remove /page suffix
+    cleanedKey = cleanedKey.replace(/\/page$/g, '');
```

Comment:
@conico974 I'm using `app/api/something/route.tsx`, but the key still contains `app/api/something/route`, should we also remove the `route`? The nextjs codebase deletes the last `/<thing>` from the key.

Reply:
Technically we should, but I don't think it matters that much. This is used only to modify react resolution so that your page use the same version of react as next. I don't think they do anything related to react on route

Reply:
There is actually a gotcha, https://nextjs.org/docs/app/building-your-application/routing/middleware#matching-paths. We should follow this execution order.
I think in order to do that we'll have to merge this with your middleware fix, i'll try to come up with something later today